File: src/editor/selection.c
#include "editor.h"

void editor_get_sel_offsets(int *start, int *end) {
    int a_off, b_off;
    if (editor.mark_y < editor.rows)
        a_off = editor.row[editor.mark_y].off + editor.mark_x;
    else
        a_off = editor.text.len;

    if (editor.cpos.y < editor.rows)
        b_off = editor.row[editor.cpos.y].off + editor.cpos.x;
    else
        b_off = editor.text.len;

    if (a_off <= b_off) {
        *start = a_off;
        *end   = b_off;
    } else {
        *start = b_off;
        *end   = a_off;
    }
}

void editor_select_start(void) {
    if (!editor.selecting) {
        editor.selecting = 1;
        editor.mark_x = editor.cpos.x;
        editor.mark_y = editor.cpos.y;
    }
}

void editor_select_clear(void) {
    editor.selecting = 0;
}

void editor_delete_selection(void) {
    int start, end;
    editor_get_sel_offsets(&start, &end);
    int len = end - start;
    if (len <= 0) { editor.selecting = 0; return; }

    pt_delete(&editor.text, start, len);
    editor_rebuild_rows();
    editor_offset_to_cursor(start, &editor.cpos.y, &editor.cpos.x);
    editor.selecting = 0;
    editor.dirty++;
}

void editor_copy(void) {
    if (!editor.selecting) return;

    int start, end;
    editor_get_sel_offsets(&start, &end);
    int len = end - start;
    if (len <= 0) return;

    free(editor.clipboard);
    editor.clipboard = malloc(len);
    editor.clipboard_len = len;
    pt_copy_range(&editor.text, start, len, editor.clipboard);

    FILE *pb = popen("pbcopy", "w");
    if (pb) {
        fwrite(editor.clipboard, 1, editor.clipboard_len, pb);
        pclose(pb);
    }

    editor_set_status_message("Copied %d bytes", len);
}

void editor_cut(void) {
    if (!editor.selecting) return;

    int start, end;
    editor_get_sel_offsets(&start, &end);
    int len = end - start;
    if (len <= 0) return;

    free(editor.clipboard);
    editor.clipboard = malloc(len);
    editor.clipboard_len = len;
    pt_copy_range(&editor.text, start, len, editor.clipboard);

    FILE *pb = popen("pbcopy", "w");
    if (pb) {
        fwrite(editor.clipboard, 1, editor.clipboard_len, pb);
        pclose(pb);
    }

    pt_delete(&editor.text, start, len);
    editor_rebuild_rows();
    editor_offset_to_cursor(start, &editor.cpos.y, &editor.cpos.x);
    editor.selecting = 0;
    editor.dirty++;
    editor_set_status_message("Cut %d bytes", len);
}

void editor_paste(void) {
    if (editor.selecting) editor_delete_selection();

    FILE *pb = popen("pbpaste", "r");
    if (pb) {
        char *buf = NULL;
        int buflen = 0, bufcap = 0;
        char tmp[1024];
        size_t n;
        while ((n = fread(tmp, 1, sizeof(tmp), pb)) > 0) {
            if (buflen + (int)n > bufcap) {
                bufcap = (buflen + (int)n) * 2;
                buf = realloc(buf, bufcap);
            }
            memcpy(&buf[buflen], tmp, n);
            buflen += (int)n;
        }
        pclose(pb);
        if (buf && buflen > 0) {
            free(editor.clipboard);
            editor.clipboard = buf;
            editor.clipboard_len = buflen;
        } else {
            free(buf);
        }
    }

    if (!editor.clipboard || editor.clipboard_len <= 0) return;

    if (editor.cpos.y == editor.rows) {
        if (editor.rows == 0) {
            editor_insert_row_meta(0, 0, 0);
        } else {
            int old_len = editor.text.len;
            pt_insert(&editor.text, old_len, "\n", 1);
            editor_insert_row_meta(editor.rows, old_len + 1, 0);
        }
    }

    int off = editor.row[editor.cpos.y].off + editor.cpos.x;
    pt_insert(&editor.text, off, editor.clipboard, editor.clipboard_len);
    editor_rebuild_rows();

    int target = off + editor.clipboard_len;
    editor_offset_to_cursor(target, &editor.cpos.y, &editor.cpos.x);

    editor.dirty++;
    editor_set_status_message("Pasted %d bytes", editor.clipboard_len);
}

File: src/editor/selection.zig
const std = @import("std");
const Editor = @import("Editor.zig");

fn getSelOffsets(self: *const Editor) ?struct { start: usize, end: usize } {
    const sel = self.selection orelse return null;

    const a_off = if (sel.mark_y < self.rows.items.len)
        self.rows.items[sel.mark_y].off + sel.mark_x
    else
        self.text.getTotalLength();

    const b_off = self.cursorOffset();

    if (a_off <= b_off) {
        return .{ .start = a_off, .end = b_off };
    } else {
        return .{ .start = b_off, .end = a_off };
    }
}

pub fn selectStart(self: *Editor) void {
    if (self.selection == null) {
        self.selection = .{
            .mark_x = self.cx,
            .mark_y = self.cy,
        };
    }
    self.markDirtyAll();
}

pub fn selectClear(self: *Editor) void {
    if (self.selection != null) self.markDirtyAll();
    self.selection = null;
}

pub fn deleteSelection(self: *Editor) !void {
    const offsets = getSelOffsets(self) orelse return;
    const len = offsets.end - offsets.start;
    if (len == 0) {
        self.selection = null;
        return;
    }

    self.breakSeq();
    const saved = self.allocator.alloc(u8, len) catch null;
    if (saved) |buf| {
        self.text.copyRange(offsets.start, buf);
        self.recordDelete(offsets.start, buf);
        self.allocator.free(buf);
    }
    try self.text.delete(offsets.start, len);
    try self.rebuildRows();
    self.setCursorFromOffset(offsets.start);
    self.selection = null;
    self.dirty += 1;
}

pub fn copy(self: *Editor) void {
    const offsets = getSelOffsets(self) orelse return;
    const len = offsets.end - offsets.start;
    if (len == 0) return;

    if (self.clipboard) |old| self.allocator.free(old);
    self.clipboard = self.allocator.alloc(u8, len) catch return;
    self.text.copyRange(offsets.start, self.clipboard.?);

    pbcopyWrite(self.clipboard.?);

    self.copy_flash_off = .{ .start = offsets.start, .end = offsets.end };
    self.copy_flash_start = std.time.milliTimestamp();

    self.setStatusMessage("Copied {d} bytes", .{len});
}

pub fn cut(self: *Editor) !void {
    const offsets = getSelOffsets(self) orelse return;
    const len = offsets.end - offsets.start;
    if (len == 0) return;

    if (self.clipboard) |old| self.allocator.free(old);
    self.clipboard = try self.allocator.alloc(u8, len);
    self.text.copyRange(offsets.start, self.clipboard.?);

    pbcopyWrite(self.clipboard.?);

    self.breakSeq();
    self.recordDelete(offsets.start, self.clipboard.?);
    try self.text.delete(offsets.start, len);
    try self.rebuildRows();
    self.setCursorFromOffset(offsets.start);
    self.selection = null;
    self.dirty += 1;
    self.setStatusMessage("Cut {d} bytes", .{len});
}

pub fn paste(self: *Editor) !void {
    if (self.selection != null) try self.deleteSelection();

    // Try to read from system clipboard via pbpaste
    const sys_clip = pbpasteRead(self.allocator);
    if (sys_clip) |clip| {
        if (self.clipboard) |old| self.allocator.free(old);
        self.clipboard = clip;
    }

    const clip = self.clipboard orelse return;
    if (clip.len == 0) return;

    if (self.cy >= self.rows.items.len) return;

    const off = self.cursorOffset();
    self.breakSeq();
    self.recordInsert(off, clip);
    try self.text.insert(off, clip);
    try self.rebuildRows();
    self.setCursorFromOffset(off + clip.len);
    self.dirty += 1;
    self.setStatusMessage("Pasted {d} bytes", .{clip.len});
}

fn pbcopyWrite(data: []const u8) void {
    var child = std.process.Child.init(&.{"pbcopy"}, std.heap.page_allocator);
    child.stdin_behavior = .Pipe;
    child.stdout_behavior = .Close;
    child.stderr_behavior = .Close;
    child.spawn() catch return;
    if (child.stdin) |*stdin| {
        stdin.writeAll(data) catch {};
        stdin.close();
        child.stdin = null;
    }
    _ = child.wait() catch {};
}

fn pbpasteRead(allocator: std.mem.Allocator) ?[]u8 {
    var child = std.process.Child.init(&.{"pbpaste"}, allocator);
    child.stdout_behavior = .Pipe;
    child.stdin_behavior = .Close;
    child.stderr_behavior = .Close;
    child.spawn() catch return null;

    const stdout = child.stdout orelse {
        _ = child.wait() catch {};
        return null;
    };
    const result = stdout.readToEndAlloc(allocator, 1024 * 1024) catch {
        _ = child.wait() catch {};
        return null;
    };
    _ = child.wait() catch {};

    if (result.len == 0) {
        allocator.free(result);
        return null;
    }
    return result;
}
